### DISCUSSION TITLE
Azure openai using bearer tokens

### DISCUSSION THREAD

**Discussion by Vkodthiv** [danielmiessler/Fabric#1178] [Q&A]
I downloaded the GO binary...checked the .env file. However getting a 401 error for Azure endpoint. I have double checked the api keys and they are valid. Is it possible to print detailed stack traces/headers for the API requests?

➜  fabric git:(main) ✗ pbpaste | ./fabric --pattern summarize
error, status code: 401, status: 401 Unauthorized, message: Access denied due to invalid subscription key or wrong API endpoint. Make sure to provide a valid key for an active subscription and use a correct regional API endpoint for your resource.

**Comment by Vkodthiv** [comment]
specifically my company only allows "Microsoft Entra ID authentication" for the azure openAI endpoint....
>>>>>Authentication tokens are included in a request as the Authorization header. The token provided must be preceded by Bearer, for example Bearer YOUR_AUTH_TOKEN. 

Adding a bearer token to the .env file and AZURE_API_KEY attribute is not working

**Reply by Vkodthiv** [reply]
I see that the client library supports  Azure AD..https://github.com/sashabaranov/go-openai/blob/c203ca001fecd40210cfcf9923ab69235c92e321/client.go#L189C3-L190C80
How do i configure AD tokens in .env file ?

`	       // OpenAI or Azure AD authentication
		req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", c.config.authToken))
`

**Reply by Vkodthiv** [reply]
i see the problem....i opened a discussion on go-openai regarding this issue. seems like DefaultAzureConfig wont allow for AzureAD. 
https://github.com/sashabaranov/go-openai/discussions/910

Meanwhile are there any workarounds ? I have a demo/POC coming up with my leadership and i cant make this work with my enterprise Azure endpoints. Hmnn.

**Comment by mseigal** [comment]
My company provided an API key and I had to figure out the api base url and the models that were supported. It was a little bit of trial and error but it was not a big issue. I got it working within a few minutes.
